### PR TITLE
fix(57348): Auto-complete on satisfies auto-imports satisfies 

### LINKED ISSUE #57348: Auto-complete on `satisfies` auto-imports `satisfies`
### 🔎 Search Terms

satisfies auto import vscode

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I (didn't) reviewed the FAQ for entries about it because @DanielRosenwasser told me to open an issue

### ⏯ Playground Link

_No response_

### 💻 Code

Create a file that has `export function satisfies() {}`. Then, in a different file type `const a = 1 sat` and hit enter (to select the "satisfies" auto-complete in vscode).

### 🙁 Actual behavior

The TS language server will add an `import { satisfies } from "..."`.

### 🙂 Expected behavior

It should not do it.

### Additional information about the issue

This is only a problem if you have a module that exports `satisfies` somewhere, but this is very common: https://www.npmjs.com/package/@types/semver does and has 21M weekly downloads.

It also happens for `as`, but  I have never seen a module exporting an `as` function so I never encountered this in practice.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #57348

### CODE REVIEW

File: src/services/completions.ts [hunk 1]
```diff
@@ -4004,7 +4005,9 @@ function getCompletionData(
                         if (!isIdentifierText(symbolName, getEmitScriptTarget(host.getCompilationSettings()))) return false;
                         if (!detailsEntryId && isStringANonContextualKeyword(symbolName)) return false;
                         if (!isTypeOnlyLocation && !importStatementCompletion && !(targetFlags & SymbolFlags.Value)) return false;
+                        if (!isTypeOnlyLocation && contextToken && isExpression(contextToken) && contains([SyntaxKind.AsKeyword, SyntaxKind.SatisfiesKeyword], stringToToken(symbolName))) return false;
```
Comment (DanielRosenwasser) [inline src/services/completions.ts:4008]: Special-casing just these two feels off - why can't we just return both the auto-import and the keyword? For example, if you have a module like this:

```ts
namespace foo {
    export function await() {

    }
}

export = foo;
```

Both the keyword and auto-import are available in other files.

Comment (a-tarasyuk) [reply src/services/completions.ts:4008]: As I understand it, this _"special-casing"_ should apply in cases where the user is in a _"casting/satisfies"_ position.

```ts
1 a/**/
1 sati/**/
// should we suggest auto-import here?
```

Comment (DanielRosenwasser) [reply src/services/completions.ts:4008]: I suppose that's fine - in which case, at the very least that you need to check for no line terminator before the position, right?

Comment (andrewbranch) [reply src/services/completions.ts:4008]: Isn’t the keyword supposed to be present and sorted above the auto-import? Here’s `async`, for example:

image

In a position like `1 satisf/**/`, it’s true that an identifier would be invalid, so we could consider blocking _all_ auto-imports there. However, I sometimes find myself wanting completions in the middle of a messy edit:

```ts
const x = someFn bla/**/, foo); // going to backtrack and add the missing `(`
```

so I’m hesitant to block completions based on relying too much on the current state of the parse tree. It feels to me like the bug is just that the keywords don’t show up here, even though they’re valid:

image

Comment (a-tarasyuk) [reply src/services/completions.ts]: Screenshot 2024-02-14 at 03 42 12
Screenshot 2024-02-14 at 03 42 26

Comment (Mateusnasciment) [reply src/services/completions.ts]: special-casing two items might seem odd. Consider returning both auto-import and keyword suggestions to enhance completion ...

**Review by andrewbranch — APPROVED** [review]

**Comment by andrewbranch** [comment]
Thanks @a-tarasyuk!

**Pull request merged**